[#59885] Cancellation dialogs for one-time and recurring meetings

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/59885

# What are you trying to accomplish?

- This PR targets the meetings module.
- Replaces standard browser `confirm` dialogs used to confirm deletion with new Primerized `DangerConfirmationDialog` component.

## Screenshots

### 1. One-time (`StructuredMeeting`)

<img width="511" alt="image" src="https://github.com/user-attachments/assets/71220967-87ca-496a-ab1c-1835dd45e859" />

**Invoked by:**

- meeting actions on meeting show page (URL `http://localhost:5000/meetings/35`)
- OR row actions on meetings index page (URL `http://localhost:5000/meetings?upcoming=false`)

<img width="287" alt="Screenshot 2025-01-30 at 17 41 05" src="https://github.com/user-attachments/assets/2012bf92-6a51-4ba7-97eb-e67f1dc38139" />
<img width="260" alt="image" src="https://github.com/user-attachments/assets/fd52d9e1-e9d5-47fa-bbfb-e4d9d4f746f6" />

### 2. Instantiated occurrences (`StructuredMeeting` with associated `RecurringMeeting`)

<img width="518" alt="image" src="https://github.com/user-attachments/assets/c7e8c138-43b0-4a2e-a571-2883dac855a0" />

**Invoked by:**

- meeting actions on meeting show page (URL `http://localhost:5000/projects/brand-new-project/meetings/32`)
- OR row actions on meetings index page (URL `http://localhost:5000/meetings`)
- OR row actions on recurring meeting show page (URL `http://localhost:5000/projects/brand-new-project/recurring_meetings/7`)

<img width="277" alt="image" src="https://github.com/user-attachments/assets/37696c1e-620b-45da-8a30-48a0ae0e272f" />
<img width="263" alt="Screenshot 2025-01-30 at 17 48 15" src="https://github.com/user-attachments/assets/04510c70-52aa-4b80-a0fc-966f9c0107b8" />
<img width="255" alt="image" src="https://github.com/user-attachments/assets/89e50c51-7847-48c8-993d-13c80a5f47b7" />

### 3. Meeting series (`RecurringMeeting`)

#### with either zero or infinite meetings in the series

<img width="513" alt="Screenshot 2025-02-03 at 13 39 31" src="https://github.com/user-attachments/assets/ec9a7b0f-51bd-4372-a887-69f058ebe18b" />

#### with one remaining meeting in the series

<img width="506" alt="Screenshot 2025-02-03 at 13 42 11" src="https://github.com/user-attachments/assets/034d8868-2e82-4d22-8e42-2a75e65ac56c" />

#### with more than one remaining meeting in the series

<img width="498" alt="Screenshot 2025-02-03 at 13 40 06" src="https://github.com/user-attachments/assets/0d761206-a0e9-4e87-bb55-af87918ca80a" />


**Invoked by:**

- actions on recurring meeting show page (URL `http://localhost:5000/projects/brand-new-project/recurring_meetings/7`)

<img width="286" alt="image" src="https://github.com/user-attachments/assets/f284e335-2a89-4b0c-abd4-26400227ea9f" />

### 4. Scheduled occurrences (`ScheduledMeeting`)

<img width="507" alt="image" src="https://github.com/user-attachments/assets/4b927abf-a5e5-4193-8da5-30fcb2e9630c" />

**Invoked by:**

- row actions on recurring meeting show page (URL `http://localhost:5000/projects/brand-new-project/recurring_meetings/7`)

<img width="264" alt="image" src="https://github.com/user-attachments/assets/b7d1e5ef-7293-4195-b16d-f92926c422e5" />


# What approach did you choose and why?

**This PR implements three DeleteDialog components, along with three accompanying `delete_*dialog` Controller actions.**  

- `Meetings::DeleteDialogComponent` (see screenshots 1 & 2 above) this renders different text depending on whether the meeting is a one-off meeting or an instantiated occurrence.
- `RecurringMeetings::DeleteDialogComponent` (see screenshot 3)
- `RecurringMeetings::DeleteScheduledDialogComponent` (see screenshot 4)

`Meetings::DeleteDialogComponent` could also have been implemented as two separate dialogs. My rule of thumb however was to stick to a one-to-one mapping with controller actions. (FWIW I also considered the opposite approach - having one or two polymorphic delete dialog implementations for the entire meetings module - but quickly discarded for reasons of complexity).

A significant part of work on this PR has been to ensure that the end user stays within either the global or project context when deleting a meeting. 

I configured routes for the `delete_*dialog` and `destroy` Controller actions within the top-level `meetings` resource and the nested `project/meetings` resource. An alternative strategy might have been to define a single global route for `delete_*dialog`/`destroy` actions and somehow pass in a redirect URL on dialog creation.

Ultimately though, I felt that creating separate routes was the simplest (and most Rails-like) strategy. That said, a large number of the tests that I wrote concern basically just sanity-check that project or global-scope is not lost - either on invoking the dialog, or on form submission.

:warning: This PR (specifically [96308af](https://github.com/opf/openproject/pull/17569/commits/96308af5121445d7b916360fe9a113c55303632c)) _may fix_ scenarios where a user is switched from global scope to project-scope. **However this needs some further manual testing.** While I've made sure the current spec suite passes, I felt that adding additional request specs/features for non-deletion actions was out-of-scope for this (already long-running) PR. _I'd recommend we create a separate work package to look into this._

This Pull Request adds basic specs for the `DangerDialog` components as well as the related `RowComponents`. Again, these specs could probably be DRY-ed up (sanity-checking project vs global scope might be something we could do with `shared_examples`).

### Danger Dialog 

There is a [draft PR to implement Danger Dialog warning variant, without second step confirmation](https://github.com/opf/primer_view_components/pull/234).

Taking into account the UX guidelines in #17789, my understanding is that the following end-user presented dialogs (referencing screenshots above) will be candidates for the simpler "warning" variant of the Danger Dialog:

1. One-time (`StructuredMeeting`)
2. Instantiated occurrences (`StructuredMeeting` with associated `RecurringMeeting`)
4. Scheduled occurrences (`ScheduledMeeting`)

We will only require the second-step confirmation check box for this dialog:

3. Meeting series (`RecurringMeeting`)

This PR can go ahead and be merged as-is, but given that there will be a forthcoming "warning" variant, I did not customise the check box label text for 1, 2 & 4. I also kept the confirmation button text as the default "Delete permanently"**, except in the case of 4, where we always refer to the action as "Cancel" and not "Delete".

@psatyal Please let me know if this looks ok to you?

** I plan to change the default button text from "Delete permanently" to "Delete" for the warning variant of the Danger Dialog / @HDinger 


# Merge checklist

- [X] Delete structured meetings
- [X] Delete recurring meetings
- [X] Delete scheduled recurring meetings
- [X] Added/updated tests
- [X] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc commented on modules/meeting/spec/features/recurring_meetings/recurring_meeting_crud_spec.rb:
```diff
@@ -89,13 +89,17 @@
+      click_on "Delete permanently"
     end
 
     expect(page).to have_current_path meetings_path # check path
```
Hi @oliverguenther. I saw you added this comment. Is that a `TODO` (i.e. do we still need to check if this is the correct path), or just an explanation of what the line does? - it seems unclear to me if we always want to forward to the global meetings path after meeting deletion

myabc replied:
`RecurringMeetinsController#destroy` redirects to `polymorphic_path([@project, :meetings])`, so either the assertion or the implementation must be wrong? 😕

oliverguenther replied:
We want to keep the current project context, if you are in one.

So if you are looking at `/projects/my-identifier/meetings/1234`, and start deleting from there, I would expect you to be redirected to `/projects/my-identifier/meetings`.

Likewise when destroying the meeting from the global meeting page `/meetings/1234`, you should be redirected to `/meetings`. It could be that we're missing the correct implementation here, or that the spec is incorrect. I will double check the current behavior

myabc commented on modules/meeting/app/controllers/recurring_meetings_controller.rb:
```diff
@@ -130,6 +132,13 @@ def update
+  def delete_dialog
+    respond_with_dialog RecurringMeetings::DeleteDialogComponent.new(
+      meeting: @recurring_meeting,
+      project: @project
```
(I based my `delete_dialog` controller action on `details_dialog`) ❓  Not sure we need to pass in project here? - see other comments on nested (project scoped) routes.

oliverguenther replied:
If you don't need to create routes depending on the current project context, then you can omit  `@project`.

myabc commented on modules/meeting/app/components/meetings/delete_dialog_component.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+
+    def recurring_meeting
+      return if recurring?
```
I'm not sure if this check is necessary (need to familiarise myself properly with `Meeting`/`StructuredMeeting`/`RecurringMeeting` domain models and their relationships)

oliverguenther replied:
It's a bit convoluted due to the existence of the old meeting.

`Meeting` is the deprecated type with two text areas stored separately from another (MeetingAgenda and MeetingMinute).

`StructuredMeeting` is the new primerized meeting with multiple primerized

`RecurringMeeting` is a meeting series, which consists of multiple `StructuredMeeting`. Every `StructuredMeeting`, which is connected through the `belongs_to :recurring_meeting`.

On top of that, there are also `ScheduledMeeting`, which is a record of a of a meeting that has been rescheduled / moved its actual date (probably not relevant for this PR), or if an occurrence has been cancelled.

myabc commented on config/locales/en.yml:
```diff
@@ -3938,6 +3939,7 @@ en:
     For more information on what the check provides, what data is needed to provide available updates, and how to disable this check, please visit <a href="%{more_info_url}">the configuration documentation</a>.
   text_own_membership_delete_confirmation: "You are about to remove some or all of your permissions and may no longer be able to edit this project after that.\nAre you sure you want to continue?"
   text_plugin_assets_writable: "Plugin assets directory writable"
+  text_permanent_delete_confirmation_checkbox_label: "I understand that this deletion cannot be reversed"
```
@HDinger I think this belongs in the global I18n strings. (Not sure though if we want to reconsider in specifying a `DangerDialogComponent` default checkbox label?)

oliverguenther replied:
I agree and I try to put generalizable strings into the main en.ml so they can be reused.

HDinger replied:
Hi @myabc 

@Kharonus added the DangerDialogs to the hierarchy CF in the scope of https://github.com/opf/openproject/pull/17717. He there added the exact same strings for the button and the checkbox label. So you can rebase your PR and remove the strings here.

myabc replied:
@HDinger great, will do!

myabc commented on modules/meeting/app/components/recurring_meetings/delete_dialog_component.html.erb:
```diff
@@ -0,0 +1,46 @@
+    id:,
+    title:,
+    form_arguments: {
+      action: polymorphic_path([@project, @meeting]),
```
This was lifted from here: https://github.com/opf/openproject/pull/17569/files#diff-d10668a4a54d66743190f51acff89d1022705eeee43b368c366e7cabf564526aL42 However there is no project-scoped `destroy` route.

oliverguenther replied:
I guess having only the global destroy makes sense, as long as you do a `redirect_back` as to not accidentally leave the project scope.

We have this issue in a few places in meeting, where some actions in a project context incorrectly redirect you to the global level. This results in you losing your project sidebar, etc.

myabc replied:
> I guess having only the global destroy makes sense, as long as you do a redirect_back as to not accidentally leave the project scope.

I ended up going with the opposite approach - have documented why in the PR description.

oliverguenther commented on modules/meeting/app/components/meetings/delete_dialog_component.rb:
```diff
@@ -0,0 +1,67 @@
+    def title = I18n.t("meeting.delete_dialog.title")
+
+    def confirmation_message
+      if recurring_meeting.present?
```
I would assume this message has three cases:

1. The meeting is a one-off StructuredMeeting (it is neither a `RecurringMeeting` series nor does it have a  `recurring_meeting` assocation).
2. The meeting is part of a meeting series. (It is a `StructuredMeeting`, but has an assocation `recurring_meeting` set. The message would contain something like "This meeting is a part of the series XYZ. This will only delete this occurrence"
4. It is a meeting series object itself (it is a `RecurringMeeting`). "Deleting XYZ series will delete all occurrences..."

myabc replied:
As I've modelled it, `Meetings::DialogComponent` will always be called with a `Meeting` or `StructuredMeeting`.

oliverguenther commented on modules/meeting/app/components/meetings/delete_dialog_component.rb:
```diff
@@ -0,0 +1,67 @@
+      end
+    end
+
+    def recurring?
```
The naming is really hard to get right. In code, recurring meeting could mean two things:

1. It is a meeting series consisting of multiple meeting occurrences
2. It is the occurrence of a meeting series

Do you have suggestions to better separate this in English? In German this works better (_Terminserie_ vs _Wiederkehrende Besprechung_).

myabc commented on modules/meeting/app/models/structured_meeting.rb:
```diff
@@ -34,6 +34,10 @@ class StructuredMeeting < Meeting
            inverse_of: :meeting
   accepts_nested_attributes_for :agenda_items
 
+  def self.model_name
+    @model_name ||= ActiveModel::Name.new(self, nil, "Meeting")
+  end
```
⚠️ This may have unintended consequences.

This ensures `polymorphic_path` resolves the correct route when passed a `StructuredMeeting` (i.e. `ActiveModel::Naming.singular_route_key(StructuredMeeting) => "meeting"`).

myabc commented on modules/meeting/app/components/meetings/delete_dialog_component.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+
+    def recurring?
+      @meeting.is_a?(RecurringMeeting)
```
This component doesn't need to be polymorphic - `@meeting` should always be a `Meeting` or a `StrucuturedMeeting`.

myabc committed:
[#59885] Show DeleteDialog for meeting occurrences

myabc committed:
[#59885] Impl. DeleteDialog for structured meetings

Implements the `DangerConfirmationDialog` for both one-time meetings and
recurring meeting occurences.

N.B. this does not change the UI for classic meetings.

myabc committed:
Spec project-scoped recurring meeting deletion

myabc committed:
[#59885] Impl. DeleteDialog for recurring meetings

Implements the primerised `DangerConfirmationDialog` for deletion of
recurring meetings.

myabc committed:
[#59885] Impl. DeleteDialog for scheduled meetings

myabc committed:
Spec global structured meeting deletion

myabc committed:
Initial spec for Meetings::RowComponent

myabc committed:
Fix missing data-test-selector attribute

myabc committed:
Fix project-scoped delete dialogs invoked from row

myabc committed:
Fix global-scoped MeetingsController actions

Always setting `@project` generates incorrect form actions and URLs on
these pages.

myabc committed:
Initial spec for RecurringMeetings::RowComponent

myabc committed:
Remove extraneous before_action for history action

There is no project-scoped route mapped for `MeetingsController#history`.
As such `params[:project_id]` will always be `nil`.

myabc committed:
Rename action to destroy_scheduled for consistency

Also configures route to use HTTP `DELETE` method.

myabc committed:
Split global recurring meeting feature specs

myabc committed:
Rename DeleteScheduledDialog param, controller ivar

Renamed for consistency with other delete dialogs.

myabc commented:
@oliverguenther I noticed that [#17762](https://github.com/opf/openproject/pull/17762)) was merged a couple days ago. It looks like we're structuring our I18n strings in slightly different ways - e.g `text_meeting_series_end_early_heading` vs `recurring_meeting/delete_dialog` with nested keys. I don't feel particularly strongly about which is best but it would be good to agree on a convention.

myabc committed:
Always assert modal is opened; Use :modal selector

Uses Capybara accessible selectors `:modal` selector and accompanying
`within_modal` helper where possible.

This allows us to use `aria-labelledby` / `aria-label` as a locator.
This also ensures that:
* the dialog is actually open.
* the dialog is `aria-modal` true.

See https://github.com/citizensadvice/capybara_accessible_selectors?tab=readme-ov-file#within_modalname-find_options-block

myabc committed:
Fix delete meeting occurrence a11y title

Makes title consistent with other delete dialog titles.

myabc committed:
Try to make Capy confirmation checking more robust

Since `Primer::Alpha::Checkbox` draws a custom checkbox element with a
`::before` pseudo-selector and disables the platform-native appearance.

This *may* cause issues with checking the confirmation check box.

myabc committed:
Fix end meeting series dialog a11y title

oliverguenther commented:
Deletion fails for meetings that are `end_after=never` 
<img width="1223" alt="Bildschirmfoto 2025-02-03 um 14 21 47" src="https://github.com/user-attachments/assets/cbf6c019-7593-4f92-ab93-baf952194007" />

oliverguenther committed:
Fix start_time of instantiated meeting

psatyal reviewed (changes requested):
Hi @myabc,


I based my review on the screenshots instead of the code (since I don't trust my ability to read code erb syntax), but they look good to me. I would only make on minor modification:

> ### One-time (`StructuredMeeting`)
> <img alt="Screenshot 2025-01-30 at 17 41 13" width="514" src="https://private-user-images.githubusercontent.com/755/408290264-b73ac53b-29a5-40b3-b55c-147734b7062e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg1NzQ0OTYsIm5iZiI6MTczODU3NDE5NiwicGF0aCI6Ii83NTUvNDA4MjkwMjY0LWI3M2FjNTNiLTI5YTUtNDBiMy1iNTVjLTE0NzczNGI3MDYyZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwM1QwOTE2MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ODdjNzdlOTQ4ZTY2ZjUwMDI5NWM4NDhlNTg0YTE5MmFmNDE5ZjlmNjYyMzI3OTg3MmNhMzI5OTkxZWMzMDkwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fICV36rilqRA3FTORtqDiB4b3oH7Bk7Tjg4KiXFckJA">
> 


For this first case (one-time, basic `StructuredMeeting`), I would change the second line to the default text so it reads:

> **Delete this meeting?**  
> This action is not reversible. Please proceed with caution.
> {checkbox} I understand that this deletion cannot be reversed/
> [Cancel] [Delete permanently]

This way, when we later go back to using the non-confirmation variant, the copy for that second line does not have to be updated.

Cheers
Parimal

myabc committed:
♾️ Fix handling meeting series that end_after never

Updates confirmation messages for end user depending on `count`.
N.B. rails-i18n does not provide built-in pluralisation rules for
`infinite` so for now we treat 0 and infinity the same.

myabc commented:
> Deletion fails for meetings that are `end_after=never` <img alt="Bildschirmfoto 2025-02-03 um 14 21 47" width="1223" src="https://private-user-images.githubusercontent.com/459462/409121774-cbf6c019-7593-4f92-ab93-baf952194007.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg1OTcwOTcsIm5iZiI6MTczODU5Njc5NywicGF0aCI6Ii80NTk0NjIvNDA5MTIxNzc0LWNiZjZjMDE5LTc1OTMtNGY5Mi1hYjkzLWJhZjk1MjE5NDAwNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwM1QxNTMzMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYzQ3Y2NiMWE0MDkwNmEzZGJmMTRkZWQ0Y2NkMDZiNDZlNWFkNGQ5ZmRlMmFmMDRjZGUwZGRhYWE2MGJiYjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BdFd_tmH3KKIikauadvKPh0_sbUfoH_Mac192t0Un1k">

@oliverguenther - thanks for pointing this out. Fixed.

oliverguenther reviewed (approved):
LGTM now 👍

myabc committed:
Update confirmation message for one-time delete

per @psatyal's comment

myabc committed:
Rename confirmation message interpolation var

Renames variabble for consistency.

myabc commented:
@psatyal Cheers for the feedback!

> I based my review on the screenshots instead of the code (since I don't trust my ability to read code erb syntax), but they look good to me. I would only make on minor modification:

No worries! - thanks for looking at the screenshots over here. I was going to cross-post to OP, but it's easier with a single source of truth.
 
> > ### One-time (`StructuredMeeting`)
> 
> For this first case (one-time, basic `StructuredMeeting`), I would change the second line to the default text so it reads:
> 
> > **Delete this meeting?**
> > This action is not reversible. Please proceed with caution.
> > {checkbox} I understand that this deletion cannot be reversed/
> > [Cancel] [Delete permanently]
> 
> This way, when we later go back to using the non-confirmation variant, the copy for that second line does not have to be updated.
> 

<img width="511" alt="image" src="https://github.com/user-attachments/assets/2880671d-5b4b-4206-b41b-32cb39560285" />

Done!

psatyal reviewed (approved):
Thanks! Looks good to me.

myabc committed:
Merge branch 'dev' into feature/59885-cancellation-dialogs-recurring-meetings

myabc committed:
Remove additional feature flags for meetings

myabc commented on modules/meeting/app/components/recurring_meetings/delete_dialog_component.rb:
```diff
@@ -0,0 +1,55 @@
+    def confirmation_message
+      t("recurring_meeting.delete_dialog.confirmation_message_html",
+        title: @recurring_meeting.title,
+        count: @recurring_meeting.remaining_occurrences&.count || 0) # actually ♾️, but we show same message as for 0
```
for @klaustopher 😄

myabc commented on modules/meeting/config/locales/en.yml:
```diff
@@ -284,6 +290,30 @@ en:
+      title: "Delete meeting series"
+      heading: "Permanently delete this meeting series?"
+      confirmation_message_html:
+        zero: >
+          Deleting <strong>%{title}</strong> will delete any meetings in this series.
+          This action is not reversible. Please proceed with caution.
```
FYI @psatyal You might also be interested in this change (I've updated the visuals above under **3. Meeting series (`RecurringMeeting`)**)

We will display a slightly different confirmation message depending on the number of remaining meetings. Because of technical constraints with Rails pluralisation rules, We're also treating zero and infinite as the same for now.

I just thought you might be interested in why conflating these two scenarios doesn't work so well in other European languages:

> with regards to translating: "any" is vaguer than "all" in English (implying there 
> might not be any at all), but in German or French you'd probably use alle or toutes for both.
> 
> _löscht alle Meetings dieser Serie_
> _supprimera toutes les réunions de cette séries_

(cross-posting from an Element developer channel that I don't think you're in)

cbliard commented on modules/meeting/spec/features/recurring_meetings/recurring_meeting_crud_spec.rb:
```diff
@@ -90,7 +90,7 @@
     show_page.delete_meeting_series
     show_page.within_modal "Delete meeting series" do
-      page.check "I understand that this deletion cannot be reversed"
+      check "I understand that this deletion cannot be reversed", allow_label_click: true
```
If all check boxes will need this, maybe it would be better to enable it globally with [`automatic_label_click`](https://www.rubydoc.info/gems/capybara/3.12.0/Capybara#configure-class_method)
```ruby
Capybara.configure do |config|
  config.automatic_label_click = true
end
```

myabc replied:
@cbliard yes, I was thinking we could try enabling globally.
